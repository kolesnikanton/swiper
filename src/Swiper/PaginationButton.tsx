import React from 'react';

type PaginationButtonT = {
  slideNumber: number,
  currentSlideNumber: number,
  onClick: (slideNumber: number) => void,
}

export default function PaginationButton({
  slideNumber, currentSlideNumber, onClick,
}: PaginationButtonT) {
  const handlePaginationButton = () => {
    onClick(slideNumber);
  };

  const isActive = slideNumber === currentSlideNumber;
  const className = isActive ? 'swiper-pagination__button swiper-pagination__button_active' : 'swiper-pagination__button';

  return (
    <button
      aria-label={`Slide ${slideNumber}`}
      id={String(slideNumber)}
      disabled={isActive}
      className={className}
      type="button"
      onClick={handlePaginationButton}
    />
  );
}
