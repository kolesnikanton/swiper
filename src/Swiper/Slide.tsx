import React from 'react';

type SlideT = {
  slide: React.ReactElement,
}

export default class Slide extends React.PureComponent<SlideT> {
  render() {
    const { slide } = this.props;
    return (
      <div className="swiper-slide">
        {slide}
      </div>
    );
  }
}
