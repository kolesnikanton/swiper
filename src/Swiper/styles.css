.swiper-container {
  margin: 0 auto;
  overflow: hidden;
}

.swiper {
  display: flex;
}

.swiper-slide {
  display: flex;
  flex-shrink: 0;
  flex-basis: 100%;

  /* To get rid of flicker */
  transform: translate3d(0, 0, 0);
}

.swiper-slide > * {
  width: 100%;

  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -o-user-select: none;
  user-select: none;
  pointer-events: none;
}

.swiper-pagination {
  display: flex;
  justify-content: center;
}

.swiper-pagination__button {
  margin: 15px 5px;
  padding: 0;
  width: 10px;
  height: 10px;
  border: none;
  border-radius: 50%;
  background: #707070;
  color: #707070;
  cursor: pointer;
  opacity: 0.4;
}

.swiper-pagination__button:hover {
  opacity: 0.6;
}

.swiper-pagination__button_active {
  opacity: 1;
}

.swiper-pagination__button:disabled:hover {
  opacity: 1;
}

.swiper-pagination__button:disabled {
  cursor: default;
}
