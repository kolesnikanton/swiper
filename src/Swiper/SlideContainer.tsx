import React from 'react';

import Slide from './Slide';

type SlideContainerT = {
  handleSwipeStart: (event: React.TouchEvent | React.MouseEvent) => void,
  handleSwipe: (event: React.TouchEvent | React.MouseEvent) => void,
  handleSwipeEnd: (event: React.TouchEvent | React.MouseEvent) => void,
  slides: Array<React.ReactElement>
}

const SlideContainer = React.forwardRef(({
  handleSwipeStart,
  handleSwipe,
  handleSwipeEnd,
  slides,
}: SlideContainerT, swiperRef: React.Ref<HTMLDivElement>) => (
  // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
  <div
    role="presentation"
    className="swiper"
    onMouseDown={handleSwipeStart}
    onTouchStart={handleSwipeStart}
    onMouseMove={handleSwipe}
    onTouchMove={handleSwipe}
    onMouseUp={handleSwipeEnd}
    onMouseOut={handleSwipeEnd}
    onTouchEnd={handleSwipeEnd}
    ref={swiperRef}
    style={{ transform: 'translate(0, 0)', transition: 'transform 0s' }}
  >
    {slides.map((slide, index) => (
      <Slide key={index} slide={slide} />
    ))}
  </div>
));

function propsAreEqual() {
  /*
    We don't need to rerender this component currently.
    Be attention when changing the parent component!
  */
  return true;
}

export default React.memo(SlideContainer, propsAreEqual);
