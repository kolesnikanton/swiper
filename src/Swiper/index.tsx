import React, { useState, useRef } from 'react';

import SlideContainer from './SlideContainer';
import PaginationButton from './PaginationButton';

type OnSwipeT = {
  currentSlide: number,
  swipedDistance: number,
  startSwipeTime: number,
  startClientX: number,
  isSwipeRight: boolean,
}

type SwiperT = {
  children: Array<React.ReactElement>,
  className?: string,
  withPagination?: boolean,
  onSwipeStart?: (
    { event, currentSlide }: { event: React.TouchEvent | React.MouseEvent } & OnSwipeT
  ) => void,
  onSwipe?: (
    { event, currentSlide }: { event: React.TouchEvent | React.MouseEvent } & OnSwipeT
  ) => void,
  onSwipeEnd?: ({ currentSlide }: OnSwipeT) => void,
  onPaginationChange?: (
    { previousSlideNumber, slideNumber }: { previousSlideNumber: number, slideNumber: number}
  ) => void,
}

export default function Swiper({
  children, className, withPagination = true, onSwipeStart, onSwipe, onSwipeEnd, onPaginationChange,
}: SwiperT) {
  const containerRef = useRef(null);
  const swiperRef = useRef(null);
  const isMovingRef = useRef(false);
  const isSwipeRightRef = useRef(true);
  const startClientXRef = useRef(0);
  const swipedDistanceRef = useRef(0);
  const startSwipeTimeRef = useRef(0);
  const currentSlideRef = useRef(1); // To optimise

  const [paginationData, setPaginationData] = useState({
    currentSlide: currentSlideRef.current,
  });

  const slidesNumber = children.length;

  const getSlideWidth = () => {
    if (!containerRef.current) {
      return 0;
    }

    const containerEl = (containerRef.current as HTMLDivElement);
    return containerEl.parentElement ? containerEl.parentElement.clientWidth : 0;
  };

  const getMaxDistanceToSwipe = () => getSlideWidth() * (children.length - 1);
  const getRoundedSlideNumber = () => Math.round(swipedDistanceRef.current / getSlideWidth());

  const isNextSlideOut = ({ distance }: { distance: number }) => (
    distance < -100 || distance > getMaxDistanceToSwipe() + 100
  );

  const getClientX = (event: React.TouchEvent | React.MouseEvent) => {
    if ('touches' in event) {
      return Math.round(event.touches[0].clientX);
    }

    return event.nativeEvent.clientX;
  };

  const roundSlide = (
    { roundedSlide, slideWidth }: { roundedSlide: number, slideWidth: number },
  ) => {
    if (!swiperRef.current) {
      return;
    }

    const roundedSlideDistance = roundedSlide * slideWidth;

    const swiperElement = swiperRef.current as HTMLDivElement;
    swiperElement.style.transition = 'transform 0.2s';
    swiperElement.style.transform = `translate(${-roundedSlideDistance}px, 0px)`;
    swipedDistanceRef.current = roundedSlideDistance;

    if (currentSlideRef.current !== roundedSlide + 1) {
      setPaginationData({ currentSlide: roundedSlide + 1 });
      currentSlideRef.current = roundedSlide + 1;
    }
  };

  const roundSlideByShortTime = ({ slideWidth }: { slideWidth: number}) => {
    let roundedSlide = isSwipeRightRef.current
      ? Math.floor(swipedDistanceRef.current / slideWidth)
      : Math.ceil(swipedDistanceRef.current / slideWidth);

    if (roundedSlide < 0) {
      roundedSlide = 0;
    }

    if (roundedSlide >= slidesNumber) {
      roundedSlide = slidesNumber - 1;
    }

    roundSlide({ roundedSlide, slideWidth });
  };

  const handleSwipeStart = (event: React.TouchEvent | React.MouseEvent) => {
    if (!swiperRef.current) {
      return;
    }

    startSwipeTimeRef.current = Date.now();
    isMovingRef.current = true;
    const clientX = getClientX(event);
    startClientXRef.current = clientX + swipedDistanceRef.current;
    (swiperRef.current as HTMLDivElement).style.transition = 'transform 0s';

    if (onSwipeStart) {
      onSwipeStart({
        event,
        currentSlide: currentSlideRef.current,
        swipedDistance: swipedDistanceRef.current,
        startSwipeTime: startSwipeTimeRef.current,
        startClientX: startClientXRef.current,
        isSwipeRight: isSwipeRightRef.current,
      });
    }
  };

  const handleSwipeEnd = () => {
    if (!isMovingRef.current) {
      return;
    }

    isMovingRef.current = false;

    const endSwipeTime = Date.now() - startSwipeTimeRef.current;
    const slideWidth = getSlideWidth();

    if (endSwipeTime < 500) {
      roundSlideByShortTime({ slideWidth });
    } else {
      const roundedSlide = getRoundedSlideNumber();
      roundSlide({ roundedSlide, slideWidth });
    }

    if (onSwipeEnd) {
      onSwipeEnd({
        currentSlide: currentSlideRef.current,
        swipedDistance: swipedDistanceRef.current,
        startSwipeTime: startSwipeTimeRef.current,
        startClientX: startClientXRef.current,
        isSwipeRight: isSwipeRightRef.current,
      });
    }
  };

  const handleSwipe = (event: React.TouchEvent | React.MouseEvent) => {
    if (!isMovingRef.current || !swiperRef.current) {
      return;
    }

    const clientX = getClientX(event);
    const distance = startClientXRef.current - clientX;

    if (isNextSlideOut({ distance })) {
      return;
    }

    isSwipeRightRef.current = swipedDistanceRef.current > distance;

    (swiperRef.current as HTMLDivElement).style.transform = `translate(${-distance}px, 0px)`;
    swipedDistanceRef.current = distance;

    if (onSwipe) {
      onSwipe({
        event,
        currentSlide: currentSlideRef.current,
        swipedDistance: swipedDistanceRef.current,
        startSwipeTime: startSwipeTimeRef.current,
        startClientX: startClientXRef.current,
        isSwipeRight: isSwipeRightRef.current,
      });
    }
  };

  const handlePaginationButton = (slideNumber: number) => {
    if (!swiperRef.current) {
      return;
    }

    const swiperElement = swiperRef.current as HTMLDivElement;
    swiperElement.style.transition = 'transform 0.8s ease-in';
    const distance = getSlideWidth() * (slideNumber - 1);
    swiperElement.style.transform = `translate(${-distance}px, 0px)`;

    if (onPaginationChange) {
      onPaginationChange({ previousSlideNumber: currentSlideRef.current, slideNumber });
    }

    swipedDistanceRef.current = distance;
    currentSlideRef.current = slideNumber;
    setPaginationData({ currentSlide: slideNumber });
  };

  const renderPagination = () => {
    const buttons = [];

    for (let i = 0; i < slidesNumber; i++) {
      const slideNumber = i + 1;

      buttons.push(
        <PaginationButton
          key={slideNumber}
          slideNumber={slideNumber}
          currentSlideNumber={paginationData.currentSlide}
          onClick={handlePaginationButton}
        />,
      );
    }

    return buttons;
  };

  const defaultClassName = 'swiper-container';
  const classNames = className ? `${defaultClassName} ${className}` : defaultClassName;

  return (
    <>
      <div ref={containerRef} className={classNames}>
        <SlideContainer
          slides={children}
          ref={swiperRef}
          handleSwipeStart={handleSwipeStart}
          handleSwipe={handleSwipe}
          handleSwipeEnd={handleSwipeEnd}
        />
      </div>
      {
        withPagination ? (
          <div className="swiper-pagination">
            {renderPagination()}
          </div>
        ) : null
      }
    </>
  );
}
