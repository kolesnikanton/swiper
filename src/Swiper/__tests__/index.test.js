import React from 'react';
import TestRenderer from 'react-test-renderer';
import Swiper from '..';

test('render', () => {
  const component = TestRenderer.create(
    <Swiper>
      <img
        alt="1"
        src="https://antonkolesnik.com/38df7b2ad210d02fbd81.jpg"
      />
      <h1 className="slide">Slide 2</h1>
      <img
        alt="3"
        src="https://antonkolesnik.com/6ab753c0e5734915a8c9.jpg"
      />
    </Swiper>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render without pagination', () => {
  const component = TestRenderer.create(
    <Swiper withPagination={false}>
      <img
        alt="1"
        src="https://antonkolesnik.com/38df7b2ad210d02fbd81.jpg"
      />
      <h1 className="slide">Slide 2</h1>
      <img
        alt="3"
        src="https://antonkolesnik.com/6ab753c0e5734915a8c9.jpg"
      />
    </Swiper>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render with additional props', () => {
  const component = TestRenderer.create(
    <Swiper
      className="custom-class"
      onSwipeStart={() => ({})}
      onSwipe={() => ({})}
      onSwipeEnd={() => ({})}
    >
      <img
        alt="1"
        src="https://antonkolesnik.com/38df7b2ad210d02fbd81.jpg"
      />
      <h1 className="slide">Slide 2</h1>
      <img
        alt="3"
        src="https://antonkolesnik.com/6ab753c0e5734915a8c9.jpg"
      />
    </Swiper>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
