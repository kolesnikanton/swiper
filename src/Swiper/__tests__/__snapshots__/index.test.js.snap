// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
[
  <div
    className="swiper-container"
  >
    <div
      className="swiper"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="presentation"
      style={
        {
          "transform": "translate(0, 0)",
          "transition": "transform 0s",
        }
      }
    >
      <div
        className="swiper-slide"
      >
        <img
          alt="1"
          src="https://antonkolesnik.com/38df7b2ad210d02fbd81.jpg"
        />
      </div>
      <div
        className="swiper-slide"
      >
        <h1
          className="slide"
        >
          Slide 2
        </h1>
      </div>
      <div
        className="swiper-slide"
      >
        <img
          alt="3"
          src="https://antonkolesnik.com/6ab753c0e5734915a8c9.jpg"
        />
      </div>
    </div>
  </div>,
  <div
    className="swiper-pagination"
  >
    <button
      aria-label="Slide 1"
      className="swiper-pagination__button swiper-pagination__button_active"
      disabled={true}
      id="1"
      onClick={[Function]}
      type="button"
    />
    <button
      aria-label="Slide 2"
      className="swiper-pagination__button"
      disabled={false}
      id="2"
      onClick={[Function]}
      type="button"
    />
    <button
      aria-label="Slide 3"
      className="swiper-pagination__button"
      disabled={false}
      id="3"
      onClick={[Function]}
      type="button"
    />
  </div>,
]
`;

exports[`render with additional props 1`] = `
[
  <div
    className="swiper-container custom-class"
  >
    <div
      className="swiper"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="presentation"
      style={
        {
          "transform": "translate(0, 0)",
          "transition": "transform 0s",
        }
      }
    >
      <div
        className="swiper-slide"
      >
        <img
          alt="1"
          src="https://antonkolesnik.com/38df7b2ad210d02fbd81.jpg"
        />
      </div>
      <div
        className="swiper-slide"
      >
        <h1
          className="slide"
        >
          Slide 2
        </h1>
      </div>
      <div
        className="swiper-slide"
      >
        <img
          alt="3"
          src="https://antonkolesnik.com/6ab753c0e5734915a8c9.jpg"
        />
      </div>
    </div>
  </div>,
  <div
    className="swiper-pagination"
  >
    <button
      aria-label="Slide 1"
      className="swiper-pagination__button swiper-pagination__button_active"
      disabled={true}
      id="1"
      onClick={[Function]}
      type="button"
    />
    <button
      aria-label="Slide 2"
      className="swiper-pagination__button"
      disabled={false}
      id="2"
      onClick={[Function]}
      type="button"
    />
    <button
      aria-label="Slide 3"
      className="swiper-pagination__button"
      disabled={false}
      id="3"
      onClick={[Function]}
      type="button"
    />
  </div>,
]
`;

exports[`render without pagination 1`] = `
<div
  className="swiper-container"
>
  <div
    className="swiper"
    onMouseDown={[Function]}
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="presentation"
    style={
      {
        "transform": "translate(0, 0)",
        "transition": "transform 0s",
      }
    }
  >
    <div
      className="swiper-slide"
    >
      <img
        alt="1"
        src="https://antonkolesnik.com/38df7b2ad210d02fbd81.jpg"
      />
    </div>
    <div
      className="swiper-slide"
    >
      <h1
        className="slide"
      >
        Slide 2
      </h1>
    </div>
    <div
      className="swiper-slide"
    >
      <img
        alt="3"
        src="https://antonkolesnik.com/6ab753c0e5734915a8c9.jpg"
      />
    </div>
  </div>
</div>
`;
